server/lib/parties/party-service.ts

```diff
@@ -0,0 +1,228 @@
+  public code: PartyServiceErrorCode
+
+  constructor(code: PartyServiceErrorCode, message: string) {
+    super(message)
+    this.code = code
+  }
+}
+
+export function getInvitesPath(partyId: string): string {
+  return `/parties/invites/${partyId}`
+}
+
+export function getPartyPath(partyId: string): string {
+  return `/parties/${partyId}`
+}
+
+@singleton()
+export default class PartyService {
+  private parties = new Map<string, PartyRecord>()
+  private clientSocketsToPartyId = new Map<ClientSocketsGroup, string>()
+
+  constructor(
+    private nydus: NydusServer,
+    private clientSocketsManager: ClientSocketsManager,
+    private userSocketsManager: UserSocketsManager,
+  ) {}
+
+  invite(leader: PartyUser, leaderClientId: string, invites: PartyUser[]): Readonly<PartyRecord> {
+    const leaderClientSockets = this.getClientSockets(leader.id, leaderClientId)
+
+    let party = this.getClientParty(leaderClientSockets)
+    if (party) {
+      if (party.leaderId !== leader.id) {
+        throw new PartyServiceError(
+          PartyServiceErrorCode.InsufficientPermissions,
+          'Only party leader can invite people',
+        )
+      }
+
+      for (const invite of invites) {
+        party.invites.set(invite.id, invite)
+      }
+
+      this.publishToParty(party.id, {
+        type: 'invite',
+        invites,
+      })
+    } else {
+      const partyId = cuid()
+      party = {
+        id: partyId,
+        invites: new Map(invites.map(i => [i.id, i])),
+        members: new Map([[leader.id, leader]]),
+        leaderId: leader.id,
+      }
+
+      this.parties.set(partyId, party)
+      this.clientSocketsToPartyId.set(leaderClientSockets, partyId)
+      this.subscribeToParty(leaderClientSockets, party)
+    }
+
+    invites
+      .map(i => this.getUserSockets(i.name))
+      .forEach(userSockets => {
+        userSockets.subscribe(
+          getInvitesPath(party!.id),
+          () => ({
+            type: 'invite',
+            from: party!.leaderId,
+          }),
+          () => {
+            // TODO(2Pac): Handle user quitting; need to keep a map of user -> invites?
+          },
+        )
+      })
+
+    return party
+  }
+
+  removeInvite(partyId: string, target: PartyUser, leader?: PartyUser) {
+    const party = this.parties.get(partyId)
+    if (!party) {
+      throw new PartyServiceError(PartyServiceErrorCode.PartyNotFound, 'Party not found')
+    }
+
+    if (leader && leader.id !== party.leaderId) {
+      throw new PartyServiceError(
+        PartyServiceErrorCode.InsufficientPermissions,
+        'Only party leaders can remove invites to other people',
+      )
+    }
+
+    // TODO(2Pac): Check if the target is invited first before trying to do any of the below?
+
+    party.invites.delete(target.id)
+    this.publishToParty(partyId, {
+      type: 'decline',
+      target,
+    })
+
+    // TODO(2Pac): Do we want to remove the party record if there's only one member left in a party
```

Comment:
I'd say probably, yes, although it depends on how much the UI changes between being in a party/not being in a party

Reply:
Yeah, initially I did write the code to remove the party record, but then I thought if that results in party chat being destroyed, then the person who invited someone won't even see the message that someone declined. Gonna leave this as TODO for now, until we settle on a UI.